---
title: "Jurkat project"
author: "Juba"
date: "`r Sys.Date()`"
output: 
  html_document:
    number_sections: no
    theme: flatly
    highlight: tango
    toc: yes
    toc_depth: 6
    toc_float:
      collapsed: false
      smooth_scroll: true
    df_print: paged 
    css: "style.css"

params:
  MaxCount_threshold:
    value: 20
    label: "MaxCount_threshold : (filter genes with a max count under this value)"
  CpmCount_threshold:
    value: 0.5
    label: "CpmCount_threshold : (count per million threshold)"
  MinSample:
    value: 3
    label: "MinSample : (samples to apply the cpm threshold)"
  logfc_thresh: 1
---

```{r setup, include=FALSE, message = FALSE, warning = FALSE, echo = FALSE}
## Chunks configuration :
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)

## Directories :
out_fig_explore <- "out_figures/exploration/"
out_fig_dea <- "out_figures/dea/"
out_obj <- "out_objects/"

## Import packages :
source("scripts/packages.R") 

## Import functions :
source("scripts/functions.R")
```


# Project description

We are studying the effect of an inhibitor of the cGAS-STING signaling pathway __H151__ on T-ALL model cell line __Jurkat__.

The biological experiment revealed that H151 causes cell death, so the pathway is important 
for survival of T-ALL.

[We want to explore the differential expressed genes between normal condition and by inhibiting the pathway. Thus we are going to perform a differential expression analysis (DEA) followed by a pathway enrichment analysis (PEA)]{.purple .bold}

__Following a basic RNA-seq pipeline analysis__



# Importing data {.tabset .tabset-fade}

```{r import-data}
## ---- Import all the data : ----
############################# /
all_data <- read_xlsx(path = "data/S22254_alldata.xlsx",
                      sheet = "S22254_alldata")

colnames(all_data) <- gsub(" ", "_", colnames(all_data))



## ---- Extract raw counts : ----
############################ /
raw_counts <- all_data[,1:7] %>% 
  column_to_rownames("Ensembl_Gene_ID") %>% 
  dplyr::rename(LXGP1 = "LXGP1_(raw)",
                LXGP2 = "LXGP2_(raw)",
                LXGP3 = "LXGP3_(raw)",
                LXGP4 = "LXGP4_(raw)",
                LXGP5 = "LXGP5_(raw)",
                LXGP6 = "LXGP6_(raw)")



## ---- Extract meta data : ----
########################### /
metadata <- all_data[,c(1,27:31)] %>% 
  column_to_rownames("Ensembl_Gene_ID") %>% 
  mutate(Gene_Length = abs(Start_gene_position - End_gene_position))



## ---- Extract experimental data : ----
################################### /
experimental_data <- read_xlsx(path = "data/Table_Experimental conditions.xlsx") %>% 
  dplyr::select(c(1,3,5,7,9)) %>% 
  drop_na() %>% 
  column_to_rownames(var = "Code") %>% 
  dplyr::rename(Concentration = `Concentration (ng/µl)`,
                Volume = `Volume (µl)`) %>% 
  head(6) %>% 
  mutate(Condition = c(rep("Jurkat_ct",3), rep("Jurkat_H151",3)),
         Replicat = rep(c("1","2","3"), times = 2))

## 4: saving objects |||
write_xlsx(raw_counts, path = paste0(out_obj, "rawcounts.xlsx"))
write_xlsx(metadata, path = paste0(out_obj, "metadata.xlsx"))
write_xlsx(experimental_data, path = paste0(out_obj, "experimental_data.xlsx"))
```

## Raw counts {.unnumbered .unlisted}

```{r ,rows.print=5}
raw_counts
```

## Metadata {.unnumbered .unlisted}

```{r ,rows.print=5}
metadata
```

## Experimental data {.unnumbered .unlisted}

```{r ,rows.print=5}
experimental_data
```




# Data exploration

First let's view the distribution of the different bio types we have in our data : 

```{r biotypes-plot, fig.width=8, fig.height=6}
metadata %>% 
  group_by(Gene_biotype) %>% 
  summarise(counts = n()) %>% 
  filter(counts > 10) %>% 
  ggplot()+
  geom_bar(aes(x= counts,
               y= Gene_biotype,
               fill = ifelse(Gene_biotype == "protein_coding", "darkcyan", "gray")),
           stat = "identity",
           show.legend = F,
           width = 0.8)+
  theme_bw()+
  labs(title = "Gene biotypes distributions",
       y = "",
       fill = "",
       x = "Counts")+
  theme(plot.title = element_text(size = 14,
                                  colour = "darkcyan",
                                  face = "bold"),
        plot.background = element_rect(colour = "whitesmoke", 
                                       fill = "whitesmoke", 
                                       linewidth = 1),
        axis.text = element_text(size = 11),
        axis.title = element_text(size = 12, 
                                  colour = "darkcyan", 
                                  face = "bold"))+
  scale_fill_manual(values = c("darkcyan"="darkcyan", "gray"="gray60"))+
  xlim(c(0,17000))+
  geom_text(aes(x= counts,
                y= Gene_biotype,
                label = counts),
            nudge_x = 700,
            size = 3.5)
```

[In the downstream analysis (DEA), we'll be focusing on the top 2 biotypes (`protein_coding` and `lncRNA`). Additional filtering will be applied :]{.p-120 .red .bold}

- `MaxCount_threshold` = __`r params$MaxCount_threshold`__ (At least 1 sample must have a read count over that value)
- `CpmCount_threshold` = __`r params$CpmCount_threshold`__ (Count per million reads threshold)
- `MinSample` = __`r params$MinSample`__ (Samples that should pass the cpm threshold)

```{r}
pre_process_results <- Pre_process(RCounts = raw_counts,
                                   Conditions = experimental_data,
                                   Metadata = metadata,
                                   MaxCount_threshold = params$MaxCount_threshold,
                                   CpmCount_threshold = params$CpmCount_threshold,
                                   MinSample = params$MinSample)

## Saving filtered raw_counts
write_xlsx(pre_process_results$raw_filt, path = paste0(out_obj, "rawcounts_filtered.xlsx"))
```

<br><br>

Now to understand the global gene expression landscape and to assess the quality control of our data, we need to perform a dimentionality reduction analysis : [Principal component analysis (PCA)]{.bold .purple}

The PCA :

- Assumes homoscedasticity (equal variance) between the samples
- Assumes roughly normal distribution

So we need to perform some data transformation first on `raw counts`.

We'll be using a [Variance Stabilizing Transformation (VST)]{.bold .purple} from the `DESeq2` package. And This will :

- Normalize for library size (`DESeq2` computes size factors using a `median-of-ratios` method)
- Use a __negative binomial__ model to estimate dispersion
- Apply a variance-stabilizing mathematical transformation

```{r pca-plot}
pre_process_results$PCA_plt +
  scale_colour_manual(values = c("Jurkat_H151"="brown", "Jurkat_ct"="black"))+
  my_general_theme()
```



# Differential expression analysis (DEA)

Now let's dive into gene expression.

```{r DESeq-object, echo=TRUE}
dds <- pre_process_results$DESeqData
dds$Condition <- relevel(dds$Condition, ref = "Jurkat_ct") ## set the control
dds <- DESeq(dds) 
```

```{r DEA-res, rows.print=10}
## Get the results
res <- results(dds, contrast = c("Condition", "Jurkat_ct", "Jurkat_H151"), alpha = 0.05) %>% 
  as.data.frame() %>% 
  merge(metadata, by = 0) %>% 
  column_to_rownames(var = "Row.names") %>% 
  dplyr::filter(!(is.na(padj)))

res %>% DT::datatable(options = list(pageLength = 10, scrollX = T))
```

```{r All-results-creation}
raw_counts_filt <- pre_process_results$raw_filt

## Norm counts
n_fact <- estimateSizeFactorsForMatrix(pre_process_results$raw_filt)
tmp <- sweep(pre_process_results$raw_filt, 2, n_fact, FUN="/")
n_counts <- list(n_counts = tmp,
                 n_counts_log2 = log2(tmp+1),
                 n_counts_log10 = log10(tmp+1))

## Colnames correction
colnames(n_counts$n_counts_log2) <- paste0(colnames(n_counts$n_counts_log2),"_log2")
colnames(n_counts$n_counts_log10) <- paste0(colnames(n_counts$n_counts_log10),"_log10")
colnames(n_counts$n_counts) <- paste0("norm_",colnames(n_counts$n_counts))

## Combining the counts
all_results <- cbind(raw_counts_filt, 
                     n_counts$n_counts, 
                     n_counts$n_counts_log2, 
                     n_counts$n_counts_log10,
                     res)

## Final result (no filter)
all_results <- all_results %>% 
  dplyr::mutate(Log_basemean = log10(baseMean),
                Mean_log2_JK_ct = apply(n_counts$n_counts_log2[rownames(res),1:3],1,mean),
                Mean_log2_JK_H151 = apply(n_counts$n_counts_log2[rownames(res),4:6],1,mean),
                Gene_name = case_when(is.na(Gene_name) ~ rownames(all_results),
                                      TRUE ~ Gene_name),
                ensembl = rownames(all_results),
                padj = padj + 1e-300,
                Rank = case_when(
                  abs(log2FoldChange) >= 3 & padj < 0.5 ~ 
                    round((sign(log2FoldChange)*3) * (-log10(padj)/10), 4),
                  abs(log2FoldChange) < 3 & padj < 0.5 ~
                    round(log2FoldChange * (-log10(padj)/10), 4),
                  padj >= 0.5 ~ 0
                )) %>% 
  dplyr::arrange(desc(Rank)) %>% 
  dplyr::mutate(Rank_idx = seq(1, nrow(all_results))) %>% 
  dplyr::distinct(Gene_name, .keep_all = T)

## Final result (filtered)
all_results_filt <- all_results %>% 
  dplyr::filter(padj < 0.05,
                abs(log2FoldChange) > params$logfc_thresh) %>% 
  dplyr::arrange(desc(log2FoldChange))

## Saving the results
write_xlsx(all_results, path = paste0(out_obj, "dea_results.xlsx"))
write_xlsx(all_results_filt, path = paste0(out_obj, "dea_results_filt.xlsx"))



rm(n_fact,tmp,n_counts,dds)
```

Let's get a quick overview of the distribution of the DEGs

- Volcano plot:

```{r Volcano-plot, fig.height=7, fig.width=8}
My_volcano(all_results_filt,
           x= "log2FoldChange", 
           y= "padj", 
           lab= "Gene_name",
           point_size= 3,
           label_genes= T,
           label_genes_xthreshold= 2,
           label_genes_ythreshold= 1e-50,
           lab_size= 3.2,
           lab_max_overlap= 10,
           lab_box.padding= 0.4,
           lab_colour = "black",
           lab_force = 1, 
           lab_force_pull = .8,
           selected_genes= NULL,
           logFC_thresh= 1.5,
           padj_thresh= 1e-10,
           col_palette= NULL,
           extend_xlim= 0.5)
```

- Heatmap:

Let's draw a heatmap of the top 20up and 20down genes

```{r heatmap-setup}
## 1- table of counts and genes selection
selected_genes <- all_results_filt[c(1:20, (nrow(all_results_filt)-19):nrow(all_results_filt)),]
selected_genes <- setNames(selected_genes$Gene_name, selected_genes$ensembl)

norm_counts <- nlog_scale_data(all_results_filt[,1:6],
                               selected_genes)

## 2- log-normalized and scaled table of counts
n_count <- norm_counts$scaled

## 3- log-normalized counts for hm annotations (unscaled)
log_counts <- norm_counts$unscaled

## 4- set up annotation parameters
Cond <- as.factor(rep(c("JK_ct", "JK_H151"), each = 3))
ColConditions <- setNames(c("purple","khaki4"), levels(Cond))

## 4.1
HAnnot <- HeatmapAnnotation(Condition = Cond,
                            col = list(Condition = ColConditions),
                            annotation_name_side = "left",
                            show_annotation_name = F,
                            show_legend = T,
                            annotation_name_gp = list(fontsize = 10,
                                                      col = "navy",
                                                      fontface = "bold"),
                            annotation_legend_param = list(grid_height = unit(0.7,"cm"),
                                                           grid_width = unit(0.6,"cm"),
                                                           labels_gp = gpar(col = "black",
                                                                            fontsize = 11,
                                                                            fontface = "bold"),
                                                           title_gp = gpar(col = "darkcyan",
                                                                           fontsize = 13,
                                                                           fontface = "bold")))

## 4.2
HAnnot2 <- HeatmapAnnotation(LogCounts = anno_boxplot(log_counts, 
                                                      axis = TRUE, 
                                                      gp = gpar(fill = "orange", col = "black"),
                                                      height = unit(1.7, "cm")),
                             annotation_name_side = "left",
                             annotation_name_gp = gpar(fontsize = 10,
                                                       col = "darkcyan",
                                                       fontface = "bold"),
                             annotation_name_rot = 90, 
                             annotation_name_offset = unit(0.2, "in"))


## 4.3
heatmap_legend_param = list(legend_height = unit(3,"cm"),
                            direction = "horizontal",
                            legend_width = unit(6,"cm"),
                            title = "Z-score",
                            title_position = "topcenter",
                            title_gp = gpar(fontsize=13, 
                                            col="darkcyan", 
                                            fontface="bold"),
                            label_gp = gpar(fontsize=10, col="black"),
                            legend_gp = gpar(fontsize=10))

```

<div class='figure'>

```{r hm-plot, fig.height=8, fig.width=7}
Heatmap(n_count,
        cluster_rows = T, 
        bottom_annotation = HAnnot,
        top_annotation = HAnnot2,
        column_title = "Heatmap", 
        column_title_gp = gpar(fontsize = 16, 
                               col = "darkcyan",
                               fontface = "bold"),
        show_row_dend = T,
        show_column_dend = F,
        column_names_gp = gpar(col = "black", 
                               fontface = "bold",
                               fontsize = 11),
        column_names_rot = 60,
        column_labels = colnames(n_count),
        row_labels = all_results_filt[names(selected_genes),"Gene_name"],
        name = "Z-score",
        cluster_columns = F,
        col = colorRamp2(c(-2,0,2), 
                         c("darkgreen","white","darkred")),
        row_dend_side = "left",
        heatmap_legend_param = heatmap_legend_param) %>% 
  ComplexHeatmap::draw(heatmap_legend_side = "bottom")
```

</div>


<br><br>

[DEGs list overview]{.p-140 .red .bold}

```{r degs}
genes_list <- list(
  all_genes = setNames(all_results_filt$Gene_name, all_results_filt$ensembl),
  up_genes = setNames(all_results_filt$Gene_name[all_results_filt$log2FoldChange > 0],
                      all_results_filt$ensembl[all_results_filt$log2FoldChange > 0]),
  down_genes = setNames(all_results_filt$Gene_name[all_results_filt$log2FoldChange < 0],
                        all_results_filt$ensembl[all_results_filt$log2FoldChange < 0])
)
```


- Up regulated genes: [`r length(genes_list$up_genes)`]{.bold .red}

```{r}
all_results_filt %>% 
  dplyr::filter(Gene_name %in% genes_list$up_genes) %>% 
  dplyr::count(Gene_biotype)
```

<br>
[1- `protein-coding` genes]{.bold}
<br>

```{r up-genes-prot, class.output="scroll-200 red"}
all_results_filt %>% 
  dplyr::filter(Gene_biotype == "protein_coding",
                log2FoldChange > 0) %>% 
  dplyr::pull(Gene_name)
```

[2- `lncRNA` genes]{.bold}
<br>

```{r up-genes-lncRNA, class.output="scroll-200 red"}
all_results_filt %>% 
  dplyr::filter(Gene_biotype == "lncRNA",
                log2FoldChange > 0) %>% 
  dplyr::pull(Gene_name)
```

<br>


- Down regulated genes: [`r length(genes_list$down_genes)`]{.bold .green}

```{r}
all_results_filt %>% 
  dplyr::filter(Gene_name %in% genes_list$down_genes) %>% 
  dplyr::count(Gene_biotype)
```

<br>
[1- `protein-coding` genes]{.bold}
<br>

```{r down-genes-prot, class.output="scroll-200 green"}
all_results_filt %>% 
  dplyr::filter(Gene_biotype == "protein_coding",
                log2FoldChange < 0) %>% 
  dplyr::pull(Gene_name)
```

[2- `lncRNA` genes]{.bold}
<br>

```{r down-genes-lncRNA, class.output="scroll-200 green"}
all_results_filt %>% 
  dplyr::filter(Gene_biotype == "lncRNA",
                log2FoldChange < 0) %>% 
  dplyr::pull(Gene_name)
```


<br>

[Now inspecting the gene biotypes]{.p-120 .bold .purple}

<br>

```{r ,fig.height=7, fig.width=8}

all_results_filt %>% 
  ggplot(aes(x= log2FoldChange, y= -log10(padj)))+
  geom_vline(xintercept = -1.5, linetype = 2, color = "black", linewidth = .4)+
  geom_vline(xintercept = 1.5, linetype = 2, color = "black", linewidth = .4)+
  geom_hline(yintercept = -log10(1e-10) , linetype = 2, color = "black", linewidth = .4)+
  geom_point(aes(colour = Gene_biotype), size = 3, alpha = 0.6)+
  labs(title = "Differential expressed genes",
       subtitle = "Highlighing genes biotype",
       caption = paste(paste0("protein coding : ", 
                              length(all_results_filt$Gene_biotype[all_results_filt$Gene_biotype == "protein_coding"])),
                       paste0("lncRNA : ",
                              length(all_results_filt$Gene_biotype[all_results_filt$Gene_biotype == "lncRNA"])),
                       sep = "\n"))+
  guides(colour = guide_legend(title = NULL,
                                 override.aes = list(size = 5)))+
  scale_color_manual(values = c("protein_coding"="darkred",
                                  "lncRNA"="black"))+
  xlim(c((min(all_results_filt[["log2FoldChange"]]) - 0.5),
         (max(all_results_filt[["log2FoldChange"]]) + 0.5)))+
  my_general_theme()+
  theme(
      legend.position = "top",
      legend.justification = "right",
      legend.margin = margin(b=.05, unit = "in")
  )

```

<br>

__Let's inspect the top up and down genes deeper !__

```{r ,fig.width=8, fig.height=10}
rbind(all_results_filt %>% 
        dplyr::filter(Gene_name %in% genes_list$up_genes) %>% 
        dplyr::arrange(desc(log2FoldChange)) %>% 
        head(25),
      all_results_filt %>% 
        dplyr::filter(Gene_name %in% genes_list$down_genes) %>% 
        dplyr::arrange(desc(log2FoldChange)) %>% 
        tail(35)) %>% 
  ggplot(aes(x= log2FoldChange, y= reorder(Gene_name, log2FoldChange)))+
  geom_bar(aes(fill = Gene_biotype),
           stat = "identity",
           width = 0.7)+
  scale_fill_manual(values = c("protein_coding"="darkred",
                               "lncRNA"="black"))+
  labs(title = "Biotypes of the top up and down genes",
       y = "")+
  my_general_theme()+
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.y = element_text(face = "bold")
  )

```



# Pathway enrichment analysis (PEA)

## Gene Ontology (GO) 

```{r}
## Params:
all_genes_protcoding <- all_results_filt %>% 
  dplyr::filter(Gene_biotype == "protein_coding") %>% 
  dplyr::pull(Gene_name)

down_genes_protcoding <- all_results_filt %>% 
  dplyr::filter(Gene_biotype == "protein_coding",
                log2FoldChange < 0) %>% 
  dplyr::pull(Gene_name)

terms_list <- list()

```


<div class="note">
<p class="p-120 p-bold p-purple p-underline">
NOTE:
</p>
<P class="p-bold">
For the downstream analysis, only `protein_coding` genes will be kept to ensure the significance of the resulting terms!
</p>
<P class="p-bold">
We will be using first all the `protein_coding` DEGs (`r length(all_genes_protcoding)`) then only Down-regulated one (`r length(down_genes_protcoding)`)
</p>
</div>

<br>


### All DEGs {.tabset .tabset-pills}

Let's inspect the pathways enriched with all the DEGs (up & down) that are `protein_coding`

#### Results {.unnamed .unlisted}

```{r}
go.obj <- enrichGO(gene = all_genes_protcoding,
                   OrgDb = "org.Hs.eg.db",
                   keyType = "SYMBOL",
                   minGSSize = 10,
                   maxGSSize = 1000,
                   readable = T,
                   ont = "BP")

saveRDS(go.obj, file = paste0(out_obj, "goObj_allDegs.rds"))

go.res <- go.obj@result %>%
  dplyr::filter(p.adjust < 0.05) %>%
  dplyr::arrange(p.adjust) %>% 
  dplyr::mutate(BgGenes = as.numeric(sub("/\\d+","",BgRatio)),
                RichFactor_ratio = paste0(Count,"/",as.numeric(sub("/\\d+","",BgRatio)))) %>% 
  dplyr::select(Description, RichFactor, p.adjust, GeneRatio, BgRatio, Count, BgGenes, RichFactor_ratio, geneID)

data.frame(ID = rownames(go.res), go.res) %>% write_xlsx(paste0(out_obj,"goResults_allDegs.xlsx"))

go.res %>% DT::datatable(options = list(pageLength = 5, scrollX = T))
```

<br>

[The top 40 most significant pathways enriched with these DEGs are]{.p-120 .bold .green}

```{r ,class.output="scroll-300 red"}
go.res %>% head(40) %>% dplyr::pull(Description)
```

<br>

[The top 40 pathway with the highest `RichFactor` (Count per pathway length ratio)]{.p-120 .bold .green}

```{r ,class.output="scroll-300 red"}
go.res %>% dplyr::arrange(desc(RichFactor)) %>%  head(40) %>% dplyr::pull(Description)
```

<br><br>

#### Dotplot {.unnamed .unlisted}


```{r ,fig.height=10, fig.width=9}
terms <- go.res %>% head(40)

go.res %>% 
  dplyr::filter(Description %in% terms$Description) %>% 
  dplyr::mutate(Description = ifelse(nchar(Description) <= 50,
                                     Description,
                                     paste0(substr(Description,1,46), "...."))) %>% 
  ggplot(aes(x= RichFactor, y= forcats::fct_reorder(Description, RichFactor)))+
  geom_segment(aes(xend= 0, yend= Description))+
  geom_point(aes(color= p.adjust, size= Count))+
  geom_text(aes(label = RichFactor_ratio), size = 3, colour = "darkred", nudge_x = 0.025)+
  scale_color_viridis_c(guide = guide_colorbar(reverse = T))+
  scale_size_continuous(range = c(2,8))+
  labs(title = paste0("Top ", nrow(terms), " significant enriched pathways"),
       y = "")+
  xlim(c(0, (max(terms$RichFactor)+0.05)))+
  my_general_theme()+
  theme(
    axis.text.y = element_text(face = "bold")
  )
```

<br><br>

#### Cnet plot {.unnamed .unlisted}

```{r ,fig.height=10, fig.width=12}

sel <- c("leukocyte activation",
         "leukocyte differentiation",
         "integrated stress response signaling",
         "regulation of hemopoiesis")

Net_plot(GOobject = go.obj,
         DEAres = all_results_filt,
         condition.name = "JK-H151",
         select.terms = sel,
         nterm = 3,
         edge.param.color = NULL,
         edge.param.width = .7,
         edge.param.alpha = .4,
         edge.legend.width = 1.9,
         edge.legend.position = "top",
         edge.legend.direction = "horizontal",
         edge.legend.layers = 1,
         edge.legend.fontsize = 12,
         node.param.term.color = "black",
         node.param.term.size = c(11,7),
         node.param.gene.logFC = T,
         node.param.gene.color = "darkcyan",
         node.param.gene.size = 5,
         node.param.gene.alpha = .9,
         label.genes = T,
         label.gene.color = c("red","purple","black"),
         label.gene.size = 3,
         label.gene.alpha = 0.9,
         label.gene.repel = T) -> p ; p$plot

```

<br>

<div class='figure'>

```{r ,fig.height=7, fig.width=9}
term_list <- list()
for(i in unique(p$netdata$Term)){
  term_list[[i]] <- p$netdata$Gene[p$netdata$Term == i]
}

Genes_comparison <- Display_Venn(Markers = term_list, 
                                 colpalette = NULL, 
                                 set.names = NULL, 
                                 set.name.size = 4,
                                 text.size = 4,
                                 Padding = 0.05)

Genes_comparison$plot
```

</div>

```{r}
terms_list[["termsFromGO.AllGenes"]] <- go.res$Description
```


<br><br>

### Down DEGs {.tabset .tabset-pills}

#### Results {.unnamed .unlisted}

```{r}
go.obj <- enrichGO(gene = down_genes_protcoding,
                   OrgDb = "org.Hs.eg.db",
                   keyType = "SYMBOL",
                   minGSSize = 10,
                   maxGSSize = 1000,
                   readable = T,
                   ont = "BP")

saveRDS(go.obj, file = paste0(out_obj, "goObj_downDegs.rds"))

go.res <- go.obj@result %>%
  dplyr::filter(p.adjust < 0.05) %>%
  dplyr::arrange(p.adjust) %>% 
  dplyr::mutate(BgGenes = as.numeric(sub("/\\d+","",BgRatio)),
                RichFactor_ratio = paste0(Count,"/",as.numeric(sub("/\\d+","",BgRatio)))) %>% 
  dplyr::select(Description, RichFactor, p.adjust, GeneRatio, BgRatio, Count, BgGenes, RichFactor_ratio, geneID)

data.frame(ID = rownames(go.res), go.res) %>% write_xlsx(paste0(out_obj,"goResults_downDegs.xlsx"))

go.res %>% DT::datatable(options = list(pageLength = 10, scrollX = T))
```

<br>

[The top 50 most significant pathways enriched with these DEGs are]{.p-120 .bold .green}

```{r ,class.output="scroll-300 blue"}
go.res %>% head(50) %>% dplyr::pull(Description)
```

<br>

[The top 50 pathway with the highest `RichFactor` (Count per pathway length ratio)]{.p-120 .bold .green}

```{r ,class.output="scroll-300 blue"}
go.res %>% dplyr::arrange(desc(RichFactor)) %>%  head(50) %>% dplyr::pull(Description)
```

<br><br>

#### Dotplot {.unnamed .unlisted}


```{r ,fig.height=11, fig.width=10}
terms <- go.res %>% head(50)

go.res %>% 
  dplyr::filter(Description %in% terms$Description) %>% 
  dplyr::mutate(Description = ifelse(nchar(Description) <= 50,
                                     Description,
                                     paste0(substr(Description,1,50), "...."))) %>% 
  ggplot(aes(x= RichFactor, y= forcats::fct_reorder(Description, RichFactor)))+
  geom_segment(aes(xend= 0, yend= Description))+
  geom_point(aes(color= p.adjust, size= Count))+
  geom_text(aes(label = RichFactor_ratio), 
            size = 3, 
            colour = "darkred", 
            nudge_x = ifelse(terms$RichFactor > max(terms$RichFactor)/3, 0.025, 0.045))+
  scale_color_viridis_c(guide = guide_colorbar(reverse = T))+
  scale_size_continuous(range = c(2,7))+
  labs(title = paste0("Top ", nrow(terms), " significant enriched pathways"),
       y = "")+
  xlim(c(0, (max(terms$RichFactor)+0.05)))+
  my_general_theme()+
  theme(
    axis.text.y = element_text(face = "bold")
  )
```

<br><br>

#### Cnet plot {.unnamed .unlisted}

```{r ,fig.height=10, fig.width=12}

sel <- c("leukocyte activation",
         "leukocyte differentiation",
         "ferroptosis",
         "regulation of hemopoiesis")

Net_plot(GOobject = go.obj,
         DEAres = all_results_filt,
         condition.name = "JK-H151",
         select.terms = sel,
         nterm = 3,
         edge.param.color = NULL,
         edge.param.width = .7,
         edge.param.alpha = .4,
         edge.legend.width = 1.9,
         edge.legend.position = "top",
         edge.legend.direction = "horizontal",
         edge.legend.layers = 1,
         edge.legend.fontsize = 12,
         node.param.term.color = "black",
         node.param.term.size = c(11,7),
         node.param.gene.logFC = T,
         node.param.gene.color = "darkcyan",
         node.param.gene.size = 5,
         node.param.gene.alpha = .9,
         label.genes = T,
         label.gene.color = c("red","purple","black"),
         label.gene.size = 3,
         label.gene.alpha = 0.9,
         label.gene.repel = T) -> p ; p$plot

```

<br>

<div class='figure'>

```{r ,fig.height=7, fig.width=8}
term_list <- list()
for(i in unique(p$netdata$Term)){
  term_list[[i]] <- p$netdata$Gene[p$netdata$Term == i]
}

Genes_comparison <- Display_Venn(Markers = term_list, 
                                 colpalette = NULL, 
                                 set.names = NULL, 
                                 set.name.size = 4,
                                 text.size = 4,
                                 Padding = 0.05)

Genes_comparison$plot
```

</div>

```{r}
terms_list[["termsFromGO.DownGenes"]] <- go.res$Description
```






## Molecular Signature Database (MSigDb)

The MSigDb is a collection of annotated gene sets. It contains 8 major collections:

- __H :__ hallmark gene sets
- __C1 :__ positional gene sets
- __C2 :__ curated gene sets
- __C3 :__ motifs gene sets
- __C4 :__ computational gene sets
- __C5 :__ GO gene sets
- __C6 :__ oncogenic signatures
- __C7 :__ immunologic signatures
- __C8 :__ cell Type Signature

```{r}
gene_sets <- msigdbr()
gene_sets <- gene_sets %>% 
  dplyr::select(gene_symbol, 
                ncbi_gene, 
                ensembl_gene, 
                gs_id, gs_name, 
                gs_collection, 
                gs_subcollection,
                gs_collection_name) %>% 
  dplyr::mutate(collection = paste0(gs_collection, " : ", gs_collection_name))

conv_genes <- bitr(down_genes_protcoding, 
                   fromType = "SYMBOL",
                   toType = "ENTREZID",
                   OrgDb = "org.Hs.eg.db")
```

```{r, class.output="scroll-500"}
table(gene_sets[,c("gs_collection_name","gs_collection")])
```


<br><br>

Let's see the results using different selections and using down regulated genes

```{r}
msigdb.obj.list <- list()
for(i in unique(gene_sets$collection)){
  msigdb.obj.list[[i]] <- enricher(gene = conv_genes$ENTREZID,
                                   TERM2GENE = gene_sets[gene_sets$collection == i,
                                             c("gs_name", "ncbi_gene")])
}

saveRDS(msigdb.obj.list, file = paste0(out_obj, "msigdb.obj.list.rds"))
```

```{r, class.output="scroll-500"}
for(i in names(msigdb.obj.list)){
  cat(sprintf("%-45s: %s", 
              i, 
              paste0(nrow(msigdb.obj.list[[i]]), " significant terms")),
      sep = "\n")
}

```

<br><br>

[- C6 : Oncogenic signatures]{.bold .p-120 .blue}

```{r}
msigdb.res <- msigdb.obj.list$`C6 : Oncogenic Signature`@result %>% 
  dplyr::filter(p.adjust < 0.05) %>% 
  dplyr::arrange(p.adjust)

msigdb.res %>% 
  datatable(options = list(scrollX = T, pageLength = 5))
```

<br><br>

[- C7 : Immunologic signatures]{.bold .p-120 .blue}

```{r}
msigdb.res <- msigdb.obj.list$`C7 : ImmuneSigDB`@result %>% 
  dplyr::filter(p.adjust < 0.05) %>% 
  dplyr::arrange(p.adjust)

msigdb.res %>% 
  datatable(options = list(scrollX = T, pageLength = 5))
```

<br><br>

[- C5 : GO Biological Process]{.bold .p-120 .blue}

```{r}
msigdb.res <- msigdb.obj.list$`C5 : GO Biological Process`@result %>% 
  dplyr::filter(p.adjust < 0.05) %>% 
  dplyr::arrange(p.adjust)

msigdb.res %>% 
  datatable(options = list(scrollX = T, pageLength = 5))
```

<br><br>

[- C8 : Cell Type Signature]{.bold .p-120 .blue}

```{r}
msigdb.res <- msigdb.obj.list$`C8 : Cell Type Signature`@result %>% 
  dplyr::filter(p.adjust < 0.05) %>% 
  dplyr::arrange(p.adjust)

msigdb.res %>% 
  datatable(options = list(scrollX = T, pageLength = 5))
```

<br><br>

[- H : Hallmark]{.bold .p-120 .blue}

```{r}
msigdb.res <- msigdb.obj.list$`H : Hallmark`@result %>% 
  dplyr::filter(p.adjust < 0.05) %>% 
  dplyr::arrange(p.adjust)

msigdb.res %>% 
  datatable(options = list(scrollX = T, pageLength = 5))
```


```{r}
genes_list[["all_protein_coding"]] <- all_genes_protcoding
genes_list[["down_protein_coding"]] <- down_genes_protcoding
genes_list[["up_protein_coding"]] <- setdiff(all_genes_protcoding,
                                              down_genes_protcoding)

rm(Pre_process, nlog_scale_data, selected_genes,
   sel, i, Cond, ColConditions, terms, term_list,
   raw_counts, pre_process_results, p, norm_counts,
   n_count, msigdb.res, log_counts, heatmap_legend_param,
   HAnnot, HAnnot2, Genes_comparison, gene_sets, 
   conv_genes, all_data, raw_counts_filt, res,
   all_genes_protcoding, down_genes_protcoding)
```










## Gene Set Enrichment Analysis (GSEA) {.tabset .tabset-pills}

__GSEA__ is a functional class scoring method that evaluates whether predefined gene sets show statistically significant, concordant differences between two biological conditions.
Unlike over-representation approaches, GSEA uses the entire ranked list of genes, avoiding the need for an arbitrary significance cutoff.

In this analysis, genes were ranked using a composite score based on both:

- the magnitude and direction of differential expression (`log2FoldChange`)
- the statistical significance (`adjusted p-value`, `padj`)

<br>

[Interpretation of the ranking score]{.bold .p-120 .red}

- High positive scores correspond to the most significant __up-regulated genes__
- High negative scores correspond to the most significant __down-regulated genes__
- Scores close to zero represent non-significant genes

<br>

[The ranking score was computed according to the following rule:]{.bold}

```{r, eval=FALSE, echo=TRUE}
ranking <- case_when(
  abs(log2FoldChange) >= 3 & padj < 0.5 ~
    round((sign(log2FoldChange) * 3) * (-log10(padj) / 10), 4),

  abs(log2FoldChange) < 3 & padj < 0.5 ~
    round(log2FoldChange * (-log10(padj) / 10), 4),

  padj >= 0.5 ~ 0
)
```

So basically 

$$Score=log2​(FC)×(−log10​(padj​))$$
Where beyond an absolute `log2FoldChange` value of 3, the ranking metric is only driven by the statistical significance `padj`. 

Genes with `padj ≥ 0.5` are assigned a score of zero, effectively positioning them in the middle of the ranked list.

[This scoring strategy allows a balanced integration of effect size and statistical confidence, making it suitable for downstream GSEA]{.bold .green}

```{r}
rankedGenes <- setNames(all_results$Rank, all_results$Gene_name)
```

<br>

For this analysis we going to use __gene sets__ from the MSigDB (as already seen above)

- __H :__ hallmark gene sets
- __C5 :__ GO gene sets
- __C6 :__ oncogenic signatures

```{r}
gmtFolder <- "D:/Projects/Jurkat/gsea-geneSets/"
gmtFiles <- list.files(gmtFolder)
gmtList <- list()
for(i in gmtFiles){
  gmtList[[str_split_i(i, ".v2025", 1)]] <- getGmtPaths(gmtFile = paste0(gmtFolder, i),
                                                        genes = names(rankedGenes))
}
```

```{r}
gseaResultsList <- list()
for(i in names(gmtList)){
  gseaResultsList[[i]] <- fgsea(pathways = gmtList[[i]],
                                stats = rankedGenes,
                                minSize = 5)
}

gseaResultsList <- lapply(gseaResultsList, function(x) {
  x <- x %>% 
    dplyr::filter(padj < 0.05, NES < 0) %>% 
    dplyr::mutate(nGenes = map_int(leadingEdge, length),
                  leadingEdgeFraction = nGenes / size) %>% 
    dplyr::arrange(padj)
})
```

```{r}
gmtList %>% saveRDS(paste0(out_obj, "gmtList.rds"))
gseaResultsList %>% saveRDS(paste0(out_obj, "gseaResList-neg.rds"))
```


```{css}
.gsea-gallery {
  position: relative;
  max-width: 100%;
  margin: auto;
}

.gsea-gallery img {
  display: none;
  width: 85%;
  margin: 0 auto;
  border-radius: 15px;
  box-shadow: 0 0 8px rgba(0, 0, 0, 0.3)
}

.gsea-gallery img.active {
  display: block;
}

.gsea-gallery .nav {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  background: rgba(0,0,0,0.3);
  border-radius: 15px;
  color: white;
  border: none;
  font-size: 22px;
  padding: 20px 10px;
  cursor: pointer;
  z-index: 10;
  
  &:hover {
    background-color: rgba(0, 0, 0, 0.6)
  }
}

.gsea-gallery .prev {
  left: 3px;
}

.gsea-gallery .next {
  right: 3px;
}

```

```{css}
.gsea-dots-wrapper {
  display: flex;
  justify-content: center;
  margin-top: 13px;
  overflow: hidden;
}

.gsea-dots {
  display: flex;
  gap: 8px;
  max-width: 160px; /* ~8 dots */
  overflow-x: auto;
  padding: 4px 2px;
  scrollbar-width: none;
}

.gsea-dots::-webkit-scrollbar {
  display: none;
}

.gsea-dot {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  background: rgba(0, 0, 0, 0.3);
  cursor: pointer;
  flex: 0 0 auto;
  transition: all 0.2s ease;
}

.gsea-dot.active {
  background: rgba(0, 0, 0, 0.8);
  transform: scale(1.4);
}

```




### Hallmarks

```{r}
gseaResultsList$h.all %>%
  datatable(options = list(scrollX = T, pageLength = 10))
```

[Below is the results represented in a dot plot]{.bold .purple .p-120}

```{r ,fig.height=6, fig.width=8}
gseaResultsList$h.all %>% 
  dplyr::mutate(pathway = if_else(nchar(pathway) <= 60,
                                  pathway,
                                  paste0(substr(pathway, 0, 59), "...."))) %>% 
  ggplot(aes(x= NES, y= forcats::fct_reorder(pathway, dplyr::desc(NES))))+
  geom_point(aes(colour = leadingEdgeFraction,
                 size = size),
             alpha = 0.8)+
  labs(y = "",
       title = "Significant pathways enriched by GSEA")+
  scale_color_gradient(low = "#55eedd", high = "#005040")+
  scale_size_continuous(range = c(2,6))+
  my_general_theme()+
  theme(
    axis.text = element_text(face = "bold",
                             size = 10),
    plot.title = element_text(size = 13),
    axis.title = element_text(size = 12),
    plot.caption = element_text(size = 10,
                                colour = "gray25"),
    legend.title = element_text(size = 12)
  )
```

<br><br>

[Now let's view the enrichment plots]{.bold .purple .p-120}

<br>

<div class="gsea-gallery">

```{r}
pathways <- gseaResultsList$h.all %>% 
  dplyr::pull(pathway)


for(pathway in pathways){
  p <- gseaEnrichPlot(gseaResults = gseaResultsList$h.all,
                      rankedGenes = rankedGenes,
                      pathwayName = pathway,
                      pathwayGenes = gmtList$h.all[[pathway]],
                      genesToHighlight = genes_list$down_protein_coding,
                      segmentsSizeScale = 0.05,
                      segmentLineWidth = 0.3,
                      lineWidth = 0.08,
                      lineColour = "purple")
  
  
  
  print(p)
}
```

<button class="nav prev">◀</button>
<button class="nav next">▶</button>
</div>




### C5: GO gene sets

```{r}
gseaResultsList$c5.go.bp %>% 
  datatable(options = list(scrollX = T, pageLength = 10))
```

[Below is the results represented in a dot plot]{.bold .purple .p-120}

```{r ,fig.height=9, fig.width=10}
gseaResultsList$c5.go.bp %>% 
  dplyr::mutate(pathway = if_else(nchar(pathway) <= 60,
                                  pathway,
                                  paste0(substr(pathway, 0, 59), "....")),
                nGenes = map_int(leadingEdge, length),
                leadingEdgeFraction = nGenes / size) %>% 
  ggplot(aes(x= NES, y= forcats::fct_reorder(pathway, dplyr::desc(NES))))+
  geom_point(aes(colour = leadingEdgeFraction,
                 size = size),
             alpha = 0.8)+
  labs(y = "",
       title = "Significant pathways enriched by GSEA")+
  scale_color_gradient(low = "#55eedd", high = "#005040")+
  scale_size_continuous(range = c(2,6))+
  my_general_theme()+
  theme(
    axis.text = element_text(face = "bold",
                             size = 9),
    plot.title = element_text(size = 12),
    axis.title = element_text(size = 11),
    plot.caption = element_text(size = 10,
                                colour = "gray25"),
    legend.title = element_text(size = 11)
  )
```

<br><br>

[Now let's view the enrichment plots]{.bold .purple .p-120}

<br>

<div class="gsea-gallery">

```{r}
pathways <- gseaResultsList$c5.go.bp %>% 
  dplyr::pull(pathway)

for(pathway in pathways){
  p <- gseaEnrichPlot(gseaResults = gseaResultsList$c5.go.bp,
                      rankedGenes = rankedGenes,
                      pathwayName = pathway,
                      pathwayGenes = gmtList$c5.go.bp[[pathway]],
                      genesToHighlight = genes_list$down_protein_coding,
                      segmentsSizeScale = 0.05,
                      segmentLineWidth = 0.3,
                      lineWidth = 0.08,
                      lineColour = "purple")
  
  
  
  print(p)
}
```

<button class="nav prev">◀</button>
<button class="nav next">▶</button>
</div>


### C6: oncogenic signatures

```{r}
gseaResultsList$c6.all %>% 
  datatable(options = list(scrollX = T, pageLength = 10))
```

[Below is the results represented in a dot plot]{.bold .purple .p-120}

```{r ,fig.height=9, fig.width=10}
gseaResultsList$c6.all %>% 
  dplyr::mutate(pathway = if_else(nchar(pathway) <= 60,
                                  pathway,
                                  paste0(substr(pathway, 0, 59), "....")),
                nGenes = map_int(leadingEdge, length),
                leadingEdgeFraction = nGenes / size) %>% 
  ggplot(aes(x= NES, y= forcats::fct_reorder(pathway, dplyr::desc(NES))))+
  geom_point(aes(colour = leadingEdgeFraction,
                 size = size),
             alpha = 0.8)+
  labs(y = "",
       title = "Significant pathways enriched by GSEA")+
  scale_color_gradient(low = "#55eedd", high = "#005040")+
  scale_size_continuous(range = c(2,6))+
  my_general_theme()+
  theme(
    axis.text = element_text(face = "bold",
                             size = 10),
    plot.title = element_text(size = 13),
    axis.title = element_text(size = 12),
    plot.caption = element_text(size = 10,
                                colour = "gray25"),
    legend.title = element_text(size = 12)
  )
```

<br><br>

[Now let's view the enrichment plots]{.bold .purple .p-120}

<br>

<div class="gsea-gallery">

```{r}
pathways <- gseaResultsList$c6.all %>% 
  dplyr::pull(pathway)

for(pathway in pathways){
  p <- gseaEnrichPlot(gseaResults = gseaResultsList$c6.all,
                      rankedGenes = rankedGenes,
                      pathwayName = pathway,
                      pathwayGenes = gmtList$c6.all[[pathway]],
                      genesToHighlight = genes_list$down_protein_coding,
                      segmentsSizeScale = 0.05,
                      segmentLineWidth = 0.3,
                      lineWidth = 0.08,
                      lineColour = "purple")
  
  
  
  print(p)
}
```

<button class="nav prev">◀</button>
<button class="nav next">▶</button>
</div>



```{js}
document.addEventListener("DOMContentLoaded", function () {
  document.querySelectorAll(".gsea-gallery").forEach(gallery => {

    const images = gallery.querySelectorAll("img");
    let index = 0;

    /* --- dots container --- */
    const dotsWrapper = document.createElement("div");
    dotsWrapper.className = "gsea-dots-wrapper";

    const dotsContainer = document.createElement("div");
    dotsContainer.className = "gsea-dots";

    dotsWrapper.appendChild(dotsContainer);
    gallery.appendChild(dotsWrapper);

    /* --- init dots --- */
    images.forEach((_, i) => {
      const dot = document.createElement("span");
      dot.className = "gsea-dot";
      dot.addEventListener("click", () => setActive(i));
      dotsContainer.appendChild(dot);
    });

    const dots = dotsContainer.querySelectorAll(".gsea-dot");

    function setActive(i) {
      images[index].classList.remove("active");
      dots[index].classList.remove("active");

      index = i;

      images[index].classList.add("active");
      dots[index].classList.add("active");

      /* auto-scroll dots */
      dots[index].scrollIntoView({
        behavior: "smooth",
        inline: "center",
        block: "nearest"
      });
    }

    /* init */
    images[0].classList.add("active");
    dots[0].classList.add("active");

    /* arrows */
    gallery.querySelector(".next").addEventListener("click", () => {
      setActive((index + 1) % images.length);
    });

    gallery.querySelector(".prev").addEventListener("click", () => {
      setActive((index - 1 + images.length) % images.length);
    });
  });
});
```





















