---
title: "Jurkat project"
author: "Juba"
date: "`r Sys.Date()`"
output: 
  html_document:
    number_sections: no
    theme: flatly
    highlight: tango
    toc: yes
    toc_depth: 6
    toc_float:
      collapsed: false
      smooth_scroll: true
    df_print: paged 
    css: "style.css"

params:
  MaxCount_threshold: 20
  CpmCount_threshold: 0.5
  MinSample: 3
  logfc_thresh: 1
---

```{css ,echo=FALSE}

```


```{r setup, include=FALSE, message = FALSE, warning = FALSE, echo = FALSE}
## Chunks configuration :
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)

## Directories :
out_fig_explore <- "out_figures/exploration/"
out_fig_dea <- "out_figures/dea/"
out_obj <- "out_objects/"

## Import packages :
source("scripts/packages.R") 

## Import functions :
source("scripts/functions.R")
```


# Project description

We are studying the effect of an inhibitor of the cGAS-STING signaling pathway __H151__ on T-ALL model cell line __Jurkat__.

The biological experiment revealed that H151 causes cell death, so the pathway is important 
for survival of T-ALL.

[We want to explore the differential expressed genes between normal condition and by inhibiting the pathway. Thus we are going to perform a differential expression analysis (DEA) followed by a pathway enrichment analysis (PEA)]{.p-purple .p-bold}

__Following a basic RNA-seq pipeline analysis__



# Importing data {.tabset .tabset-fade}

```{r import-data}
## ---- Import all the data : ----
############################# /
all_data <- read_xlsx(path = "data/S22254_alldata.xlsx",
                      sheet = "S22254_alldata")

colnames(all_data) <- gsub(" ", "_", colnames(all_data))



## ---- Extract raw counts : ----
############################ /
raw_counts <- all_data[,1:7] %>% 
  column_to_rownames("Ensembl_Gene_ID") %>% 
  dplyr::rename(LXGP1 = "LXGP1_(raw)",
                LXGP2 = "LXGP2_(raw)",
                LXGP3 = "LXGP3_(raw)",
                LXGP4 = "LXGP4_(raw)",
                LXGP5 = "LXGP5_(raw)",
                LXGP6 = "LXGP6_(raw)")



## ---- Extract meta data : ----
########################### /
metadata <- all_data[,c(1,27:31)] %>% 
  column_to_rownames("Ensembl_Gene_ID") %>% 
  mutate(Gene_Length = abs(Start_gene_position - End_gene_position))



## ---- Extract experimental data : ----
################################### /
experimental_data <- read_xlsx(path = "data/Table_Experimental conditions.xlsx") %>% 
  select(c(1,3,5,7,9)) %>% 
  drop_na() %>% 
  column_to_rownames(var = "Code") %>% 
  dplyr::rename(Concentration = `Concentration (ng/µl)`,
                Volume = `Volume (µl)`) %>% 
  head(6) %>% 
  mutate(Condition = c(rep("Jurkat_ct",3), rep("Jurkat_H151",3)),
         Replicat = rep(c("1","2","3"), times = 2))

## 4: saving objects |||
write_xlsx(raw_counts, path = paste0(out_obj, "rawcounts.xlsx"))
write_xlsx(metadata, path = paste0(out_obj, "metadata.xlsx"))
write_xlsx(experimental_data, path = paste0(out_obj, "experimental_data.xlsx"))
```

## Raw counts {.unnumbered .unlisted}

```{r ,rows.print=5}
raw_counts
```

## Metadata {.unnumbered .unlisted}

```{r ,rows.print=5}
metadata
```

## Experimental data {.unnumbered .unlisted}

```{r ,rows.print=5}
experimental_data
```




# Data exploration

First let's view the distribution of the different bio types we have in our data : 

```{r biotypes-plot, fig.width=8, fig.height=6}
metadata %>% 
  group_by(Gene_biotype) %>% 
  summarise(counts = n()) %>% 
  filter(counts > 10) %>% 
  ggplot()+
  geom_bar(aes(x= counts,
               y= Gene_biotype,
               fill = ifelse(Gene_biotype == "protein_coding", "darkcyan", "gray")),
           stat = "identity",
           show.legend = F,
           width = 0.8)+
  theme_bw()+
  labs(title = "Gene biotypes distributions",
       y = "",
       fill = "",
       x = "Counts")+
  theme(plot.title = element_text(size = 14,
                                  colour = "darkcyan",
                                  face = "bold"),
        plot.background = element_rect(colour = "whitesmoke", 
                                       fill = "whitesmoke", 
                                       linewidth = 1),
        axis.text = element_text(size = 11),
        axis.title = element_text(size = 12, 
                                  colour = "darkcyan", 
                                  face = "bold"))+
  scale_fill_manual(values = c("darkcyan"="darkcyan", "gray"="gray60"))+
  xlim(c(0,17000))+
  geom_text(aes(x= counts,
                y= Gene_biotype,
                label = counts),
            nudge_x = 700,
            size = 3.5)
```

[In the downstream analysis (DEA), we'll be focusing on the top 2 biotypes (`protein_coding` and `lncRNA`). Additional filtering will be applied :]{.p-120 .p-red .p-bold}

- `MaxCount_threshold` = __`r params$MaxCount_threshold`__ (At least 1 sample must have a read count over that value)
- `CpmCount_threshold` = __`r params$CpmCount_threshold`__ (Count per million reads threshold)
- `MinSample` = __`r params$MinSample`__ (Samples that should pass the cpm threshold)

```{r}
pre_process_results <- Pre_process(RCounts = raw_counts,
                                   Conditions = experimental_data,
                                   Metadata = metadata,
                                   MaxCount_threshold = params$MaxCount_threshold,
                                   CpmCount_threshold = params$CpmCount_threshold,
                                   MinSample = params$MinSample)

## Saving filtered raw_counts
write_xlsx(pre_process_results$raw_filt, path = paste0(out_obj, "rawcounts_filtered.xlsx"))
```

<br><br>

Now to understand the global gene expression landscape and to assess the quality control of our data, we need to perform a dimentionality reduction analysis : [Principal component analysis (PCA)]{.p-bold .p-purple}

The PCA :

- Assumes homoscedasticity (equal variance) between the samples
- Assumes roughly normal distribution

So we need to perform some data transformation first on `raw counts`.

We'll be using a [Variance Stabilizing Transformation (VST)]{.p-bold .p-purple} from the `DESeq2` package. And This will :

- Normalize for library size (`DESeq2` computes size factors using a `median-of-ratios` method)
- Use a __negative binomial__ model to estimate dispersion
- Apply a variance-stabilizing mathematical transformation

```{r pca-plot}
pre_process_results$PCA_plt +
  scale_colour_manual(values = c("Jurkat_H151"="brown", "Jurkat_ct"="black"))+
  my_scatter_theme()
```



# Differential expression analysis (DEA)

Now let's dive into gene expression.

```{r DESeq-object, echo=TRUE}
dds <- pre_process_results$DESeqData
dds$Condition <- relevel(dds$Condition, ref = "Jurkat_ct") ## set the control
dds <- DESeq(dds) 
```

```{r DEA-res, rows.print=10}
## Get the results
res <- results(dds, contrast = c("Condition", "Jurkat_ct", "Jurkat_H151"), alpha = 0.05) %>% 
  as.data.frame() %>% 
  merge(metadata, by = 0) %>% 
  column_to_rownames(var = "Row.names") %>% 
  dplyr::filter(!(is.na(padj)))

res
```

```{r All-results-creation}
## Norm counts
n_fact <- estimateSizeFactorsForMatrix(pre_process_results$raw_filt)
tmp <- sweep(pre_process_results$raw_filt, 2, n_fact, FUN="/")
n_counts <- list(n_counts = tmp,
                 n_counts_log2 = log2(tmp+1),
                 n_counts_log10 = log10(tmp+1))

## Colnames correction
colnames(n_counts$n_counts_log2) <- paste0(colnames(n_counts$n_counts_log2),"_log2")
colnames(n_counts$n_counts_log10) <- paste0(colnames(n_counts$n_counts_log10),"_log10")
colnames(n_counts$n_counts) <- paste0("norm_",colnames(n_counts$n_counts))

## Combining the counts
all_results <- cbind(pre_process_results$raw_filt, 
                     n_counts$n_counts, 
                     n_counts$n_counts_log2, 
                     n_counts$n_counts_log10,
                     res)

## Final result (no filter)
all_results <- all_results %>% 
  dplyr::mutate(Log_basemean = log10(baseMean),
                Rank = case_when(padj < 1e-100 ~
                                   round(log2FoldChange*(-log10(1e-100))/30, 4),
                                 padj >= 1e-100 & padj < 1e-50 ~
                                   round(log2FoldChange*(-log10(1e-70))/30, 4),
                                 padj >= 1e-50 & padj < 1e-20 ~
                                   round(log2FoldChange*(-log10(1e-30))/30, 4),
                                 padj >= 1e-20 & padj < 1e-10 ~
                                   round(log2FoldChange*(-log10(1e-15))/30, 4),
                                 padj >= 1e-10 & padj < 1e-05 ~
                                   round(log2FoldChange*(-log10(1e-07))/30, 4),
                                 padj >= 1e-05 & padj < 0.3 ~
                                   round(log2FoldChange*(-log10(padj))/30, 4),
                                 padj >= 0.3 ~ 0)) %>% 
  dplyr::arrange(desc(Rank)) %>% 
  dplyr::mutate(Rank_idx = seq(1, nrow(all_results)))

## Final result (filtered)
all_results_filt <- all_results %>% 
  dplyr::filter(padj < 0.05,
                abs(log2FoldChange) > params$logfc_thresh)

## Saving the results
write_xlsx(all_results, path = paste0(out_obj, "dea_results.xlsx"))
write_xlsx(all_results_filt, path = paste0(out_obj, "dea_results_filt.xlsx"))

rm(n_fact,tmp)
```

Let's get a quick overview of the distribution of the DEGs

```{r Volcano-plot, fig.height=6}
My_volcano(all_results_filt,
           x= "log2FoldChange", 
           y= "padj", 
           lab= "Gene_name",
           point_size= 3,
           label_genes= T,
           lab_size= 3,
           lab_max_overlap= 15,
           lab_box.padding= 0.4,
           lab_colour = "navy", 
           lab_fill = "white", 
           lab_force_pull = .5,
           selected_genes= NULL,
           logFC_thresh= 1.5,
           padj_thresh= 1e-10,
           col_palette= NULL,
           extend_xlim= 0.5)+
  theme(
    plot.background = element_rect(colour = "whitesmoke", fill = "whitesmoke"),
    legend.background = element_rect(colour = "whitesmoke", fill = "whitesmoke"),
    legend.key = element_rect(colour = "whitesmoke", fill = "whitesmoke"),
    plot.title = element_text(colour = "darkcyan", face = "bold", margin = margin(b=0.15, unit = "in"), size = 14),
    axis.title = element_text(colour = "darkcyan", face = "bold", size = 12),
    axis.text = element_text(size = 11),
    legend.text = element_text(size = 12)
  )
```








