---
title: "Jurkat project"
author: "Juba"
date: "`r Sys.Date()`"
output: 
  html_document:
    number_sections: no
    theme: flatly
    highlight: tango
    toc: yes
    toc_depth: 6
    toc_float:
      collapsed: false
      smooth_scroll: true
    df_print: paged 
    css: "style.css"

params:
  MaxCount_threshold:
    value: 20
    label: "MaxCount_threshold : (filter genes with a max count under this value)"
  CpmCount_threshold:
    value: 0.5
    label: "CpmCount_threshold : (count per million threshold)"
  MinSample:
    value: 3
    label: "MinSample : (samples to apply the cpm threshold)"
  logfc_thresh: 1
---

```{r setup, include=FALSE, message = FALSE, warning = FALSE, echo = FALSE}
## Chunks configuration :
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)

## Directories :
out_fig_explore <- "out_figures/exploration/"
out_fig_dea <- "out_figures/dea/"
out_obj <- "out_objects/"

## Import packages :
source("scripts/packages.R") 

## Import functions :
source("scripts/functions.R")
```


# Project description

We are studying the effect of an inhibitor of the cGAS-STING signaling pathway __H151__ on T-ALL model cell line __Jurkat__.

The biological experiment revealed that H151 causes cell death, so the pathway is important 
for survival of T-ALL.

[We want to explore the differential expressed genes between normal condition and by inhibiting the pathway. Thus we are going to perform a differential expression analysis (DEA) followed by a pathway enrichment analysis (PEA)]{.p-purple .p-bold}

__Following a basic RNA-seq pipeline analysis__



# Importing data {.tabset .tabset-fade}

```{r import-data}
## ---- Import all the data : ----
############################# /
all_data <- read_xlsx(path = "data/S22254_alldata.xlsx",
                      sheet = "S22254_alldata")

colnames(all_data) <- gsub(" ", "_", colnames(all_data))



## ---- Extract raw counts : ----
############################ /
raw_counts <- all_data[,1:7] %>% 
  column_to_rownames("Ensembl_Gene_ID") %>% 
  dplyr::rename(LXGP1 = "LXGP1_(raw)",
                LXGP2 = "LXGP2_(raw)",
                LXGP3 = "LXGP3_(raw)",
                LXGP4 = "LXGP4_(raw)",
                LXGP5 = "LXGP5_(raw)",
                LXGP6 = "LXGP6_(raw)")



## ---- Extract meta data : ----
########################### /
metadata <- all_data[,c(1,27:31)] %>% 
  column_to_rownames("Ensembl_Gene_ID") %>% 
  mutate(Gene_Length = abs(Start_gene_position - End_gene_position))



## ---- Extract experimental data : ----
################################### /
experimental_data <- read_xlsx(path = "data/Table_Experimental conditions.xlsx") %>% 
  dplyr::select(c(1,3,5,7,9)) %>% 
  drop_na() %>% 
  column_to_rownames(var = "Code") %>% 
  dplyr::rename(Concentration = `Concentration (ng/µl)`,
                Volume = `Volume (µl)`) %>% 
  head(6) %>% 
  mutate(Condition = c(rep("Jurkat_ct",3), rep("Jurkat_H151",3)),
         Replicat = rep(c("1","2","3"), times = 2))

## 4: saving objects |||
write_xlsx(raw_counts, path = paste0(out_obj, "rawcounts.xlsx"))
write_xlsx(metadata, path = paste0(out_obj, "metadata.xlsx"))
write_xlsx(experimental_data, path = paste0(out_obj, "experimental_data.xlsx"))
```

## Raw counts {.unnumbered .unlisted}

```{r ,rows.print=5}
raw_counts
```

## Metadata {.unnumbered .unlisted}

```{r ,rows.print=5}
metadata
```

## Experimental data {.unnumbered .unlisted}

```{r ,rows.print=5}
experimental_data
```




# Data exploration

First let's view the distribution of the different bio types we have in our data : 

```{r biotypes-plot, fig.width=8, fig.height=6}
metadata %>% 
  group_by(Gene_biotype) %>% 
  summarise(counts = n()) %>% 
  filter(counts > 10) %>% 
  ggplot()+
  geom_bar(aes(x= counts,
               y= Gene_biotype,
               fill = ifelse(Gene_biotype == "protein_coding", "darkcyan", "gray")),
           stat = "identity",
           show.legend = F,
           width = 0.8)+
  theme_bw()+
  labs(title = "Gene biotypes distributions",
       y = "",
       fill = "",
       x = "Counts")+
  theme(plot.title = element_text(size = 14,
                                  colour = "darkcyan",
                                  face = "bold"),
        plot.background = element_rect(colour = "whitesmoke", 
                                       fill = "whitesmoke", 
                                       linewidth = 1),
        axis.text = element_text(size = 11),
        axis.title = element_text(size = 12, 
                                  colour = "darkcyan", 
                                  face = "bold"))+
  scale_fill_manual(values = c("darkcyan"="darkcyan", "gray"="gray60"))+
  xlim(c(0,17000))+
  geom_text(aes(x= counts,
                y= Gene_biotype,
                label = counts),
            nudge_x = 700,
            size = 3.5)
```

[In the downstream analysis (DEA), we'll be focusing on the top 2 biotypes (`protein_coding` and `lncRNA`). Additional filtering will be applied :]{.p-120 .p-red .p-bold}

- `MaxCount_threshold` = __`r params$MaxCount_threshold`__ (At least 1 sample must have a read count over that value)
- `CpmCount_threshold` = __`r params$CpmCount_threshold`__ (Count per million reads threshold)
- `MinSample` = __`r params$MinSample`__ (Samples that should pass the cpm threshold)

```{r}
pre_process_results <- Pre_process(RCounts = raw_counts,
                                   Conditions = experimental_data,
                                   Metadata = metadata,
                                   MaxCount_threshold = params$MaxCount_threshold,
                                   CpmCount_threshold = params$CpmCount_threshold,
                                   MinSample = params$MinSample)

## Saving filtered raw_counts
write_xlsx(pre_process_results$raw_filt, path = paste0(out_obj, "rawcounts_filtered.xlsx"))
```

<br><br>

Now to understand the global gene expression landscape and to assess the quality control of our data, we need to perform a dimentionality reduction analysis : [Principal component analysis (PCA)]{.p-bold .p-purple}

The PCA :

- Assumes homoscedasticity (equal variance) between the samples
- Assumes roughly normal distribution

So we need to perform some data transformation first on `raw counts`.

We'll be using a [Variance Stabilizing Transformation (VST)]{.p-bold .p-purple} from the `DESeq2` package. And This will :

- Normalize for library size (`DESeq2` computes size factors using a `median-of-ratios` method)
- Use a __negative binomial__ model to estimate dispersion
- Apply a variance-stabilizing mathematical transformation

```{r pca-plot}
pre_process_results$PCA_plt +
  scale_colour_manual(values = c("Jurkat_H151"="brown", "Jurkat_ct"="black"))+
  my_general_theme()
```



# Differential expression analysis (DEA)

Now let's dive into gene expression.

```{r DESeq-object, echo=TRUE}
dds <- pre_process_results$DESeqData
dds$Condition <- relevel(dds$Condition, ref = "Jurkat_ct") ## set the control
dds <- DESeq(dds) 
```

```{r DEA-res, rows.print=10}
## Get the results
res <- results(dds, contrast = c("Condition", "Jurkat_ct", "Jurkat_H151"), alpha = 0.05) %>% 
  as.data.frame() %>% 
  merge(metadata, by = 0) %>% 
  column_to_rownames(var = "Row.names") %>% 
  dplyr::filter(!(is.na(padj)))

res %>% DT::datatable(options = list(pageLength = 10, scrollX = T))
```

```{r All-results-creation}
raw_counts_filt <- pre_process_results$raw_filt

## Norm counts
n_fact <- estimateSizeFactorsForMatrix(pre_process_results$raw_filt)
tmp <- sweep(pre_process_results$raw_filt, 2, n_fact, FUN="/")
n_counts <- list(n_counts = tmp,
                 n_counts_log2 = log2(tmp+1),
                 n_counts_log10 = log10(tmp+1))

## Colnames correction
colnames(n_counts$n_counts_log2) <- paste0(colnames(n_counts$n_counts_log2),"_log2")
colnames(n_counts$n_counts_log10) <- paste0(colnames(n_counts$n_counts_log10),"_log10")
colnames(n_counts$n_counts) <- paste0("norm_",colnames(n_counts$n_counts))

## Combining the counts
all_results <- cbind(raw_counts_filt, 
                     n_counts$n_counts, 
                     n_counts$n_counts_log2, 
                     n_counts$n_counts_log10,
                     res)

## Final result (no filter)
all_results <- all_results %>% 
  dplyr::mutate(Log_basemean = log10(baseMean),
                Mean_log2_JK_ct = apply(n_counts$n_counts_log2[rownames(res),1:3],1,mean),
                Mean_log2_JK_H151 = apply(n_counts$n_counts_log2[rownames(res),4:6],1,mean),
                Gene_name = case_when(is.na(Gene_name) ~ rownames(all_results),
                                      TRUE ~ Gene_name),
                ensembl = rownames(all_results),
                Rank = case_when(padj < 1e-100 ~
                                   round(log2FoldChange*(-log10(1e-100))/30, 4),
                                 padj >= 1e-100 & padj < 1e-50 ~
                                   round(log2FoldChange*(-log10(1e-70))/30, 4),
                                 padj >= 1e-50 & padj < 1e-20 ~
                                   round(log2FoldChange*(-log10(1e-30))/30, 4),
                                 padj >= 1e-20 & padj < 1e-10 ~
                                   round(log2FoldChange*(-log10(1e-15))/30, 4),
                                 padj >= 1e-10 & padj < 1e-05 ~
                                   round(log2FoldChange*(-log10(1e-07))/30, 4),
                                 padj >= 1e-05 & padj < 0.3 ~
                                   round(log2FoldChange*(-log10(padj))/30, 4),
                                 padj >= 0.3 ~ 0)) %>% 
  dplyr::arrange(desc(Rank)) %>% 
  dplyr::mutate(Rank_idx = seq(1, nrow(all_results)))

## Final result (filtered)
all_results_filt <- all_results %>% 
  dplyr::filter(padj < 0.05,
                abs(log2FoldChange) > params$logfc_thresh) %>% 
  dplyr::arrange(desc(log2FoldChange))

## Saving the results
write_xlsx(all_results, path = paste0(out_obj, "dea_results.xlsx"))
write_xlsx(all_results_filt, path = paste0(out_obj, "dea_results_filt.xlsx"))



rm(n_fact,tmp,n_counts,dds)
```

Let's get a quick overview of the distribution of the DEGs

- Volcano plot:

```{r Volcano-plot, fig.height=7, fig.width=8}
My_volcano(all_results_filt,
           x= "log2FoldChange", 
           y= "padj", 
           lab= "Gene_name",
           point_size= 3,
           label_genes= T,
           label_genes_xthreshold= 2,
           label_genes_ythreshold= 1e-50,
           lab_size= 3.2,
           lab_max_overlap= 10,
           lab_box.padding= 0.4,
           lab_colour = "black",
           lab_force = 1, 
           lab_force_pull = .8,
           selected_genes= NULL,
           logFC_thresh= 1.5,
           padj_thresh= 1e-10,
           col_palette= NULL,
           extend_xlim= 0.5)
```

- Heatmap:

```{r heatmap-setup}
## 1- table of counts and genes selection
selected_genes <- all_results_filt[c(1:20, (nrow(all_results_filt)-19):nrow(all_results_filt)),]
selected_genes <- setNames(selected_genes$Gene_name, selected_genes$ensembl)

norm_counts <- nlog_scale_data(all_results_filt[,1:6],
                               selected_genes)

## 2- log-normalized and scaled table of counts
n_count <- norm_counts$scaled

## 3- log-normalized counts for hm annotations (unscaled)
log_counts <- norm_counts$unscaled

## 4- set up annotation parameters
Cond <- as.factor(rep(c("JK_ct", "JK_H151"), each = 3))
ColConditions <- setNames(c("purple","khaki4"), levels(Cond))

## 4.1
HAnnot <- HeatmapAnnotation(Condition = Cond,
                            col = list(Condition = ColConditions),
                            annotation_name_side = "left",
                            show_annotation_name = F,
                            show_legend = T,
                            annotation_name_gp = list(fontsize = 10,
                                                      col = "navy",
                                                      fontface = "bold"),
                            annotation_legend_param = list(grid_height = unit(0.7,"cm"),
                                                           grid_width = unit(0.6,"cm"),
                                                           labels_gp = gpar(col = "black",
                                                                            fontsize = 11,
                                                                            fontface = "bold"),
                                                           title_gp = gpar(col = "darkcyan",
                                                                           fontsize = 13,
                                                                           fontface = "bold")))

## 4.2
HAnnot2 <- HeatmapAnnotation(LogCounts = anno_boxplot(log_counts, 
                                                      axis = TRUE, 
                                                      gp = gpar(fill = "orange", col = "black"),
                                                      height = unit(2, "cm")),
                             annotation_name_side = "left",
                             annotation_name_gp = gpar(fontsize = 10,
                                                       col = "darkcyan",
                                                       fontface = "bold"),
                             annotation_name_rot = 90, 
                             annotation_name_offset = unit(0.2, "in"))


## 4.3
heatmap_legend_param = list(legend_height = unit(3,"cm"),
                            direction = "horizontal",
                            legend_width = unit(6,"cm"),
                            title = "Z-score",
                            title_position = "topcenter",
                            title_gp = gpar(fontsize=13, 
                                            col="darkcyan", 
                                            fontface="bold"),
                            label_gp = gpar(fontsize=10, col="black"),
                            legend_gp = gpar(fontsize=10))

```

```{r hm-plot, fig.height=8, fig.width=7}
Heatmap(n_count,
        cluster_rows = T, 
        bottom_annotation = HAnnot,
        top_annotation = HAnnot2,
        column_title = "Heatmap", 
        column_title_gp = gpar(fontsize = 16, 
                               col = "darkcyan",
                               fontface = "bold"),
        show_row_dend = T,
        show_column_dend = F,
        column_names_gp = gpar(col = "black", 
                               fontface = "bold",
                               fontsize = 11),
        column_names_rot = 60,
        column_labels = colnames(n_count),
        row_labels = all_results_filt[names(selected_genes),"Gene_name"],
        name = "Z-score",
        cluster_columns = F,
        col = colorRamp2(c(-2,0,2), 
                         c("darkgreen","white","darkred")),
        row_dend_side = "left",
        heatmap_legend_param = heatmap_legend_param) %>% 
  ComplexHeatmap::draw(heatmap_legend_side = "bottom")
```


<br><br>

[DEGs list overview]{.p-140 .p-red .p-bold}

```{r degs}
genes_list <- list(
  all_genes = setNames(all_results_filt$Gene_name, all_results_filt$ensembl),
  up_genes = setNames(all_results_filt$Gene_name[all_results_filt$log2FoldChange > 0],
                      all_results_filt$ensembl[all_results_filt$log2FoldChange > 0]),
  down_genes = setNames(all_results_filt$Gene_name[all_results_filt$log2FoldChange < 0],
                        all_results_filt$ensembl[all_results_filt$log2FoldChange < 0])
)
```


- Up regulated genes: [`r length(genes_list$up_genes)`]{.p-bold .p-red}

```{r up-genes, class.output="scroll-200-red"}
unname(genes_list$up_genes)
```

```{r}
all_results_filt %>% 
  dplyr::filter(Gene_name %in% genes_list$up_genes) %>% 
  dplyr::count(Gene_biotype)
```



- Down regulated genes: [`r length(genes_list$down_genes)`]{.p-bold .p-green}

```{r down-genes, class.output="scroll-200-green"}
unname(genes_list$down_genes)
```

```{r}
all_results_filt %>% 
  dplyr::filter(Gene_name %in% genes_list$down_genes) %>% 
  dplyr::count(Gene_biotype)
```

<br>

[Now inspecting the gene biotypes]{.p-120 .p-bold .p-purple}

<br>

```{r ,fig.height=7, fig.width=8}

all_results_filt %>% 
  ggplot(aes(x= log2FoldChange, y= -log10(padj)))+
  geom_vline(xintercept = -1.5, linetype = 2, color = "black", linewidth = .4)+
  geom_vline(xintercept = 1.5, linetype = 2, color = "black", linewidth = .4)+
  geom_hline(yintercept = -log10(1e-10) , linetype = 2, color = "black", linewidth = .4)+
  geom_point(aes(colour = Gene_biotype), size = 3, alpha = 0.6)+
  labs(title = "Differential expressed genes",
       subtitle = "Highlighing genes biotype",
       caption = paste(paste0("protein coding : ", 
                              length(all_results_filt$Gene_biotype[all_results_filt$Gene_biotype == "protein_coding"])),
                       paste0("lncRNA : ",
                              length(all_results_filt$Gene_biotype[all_results_filt$Gene_biotype == "lncRNA"])),
                       sep = "\n"))+
  guides(colour = guide_legend(title = NULL,
                                 override.aes = list(size = 5)))+
  scale_color_manual(values = c("protein_coding"="darkred",
                                  "lncRNA"="black"))+
  xlim(c((min(all_results_filt[["log2FoldChange"]]) - 0.5),
         (max(all_results_filt[["log2FoldChange"]]) + 0.5)))+
  my_general_theme()+
  theme(
      legend.position = "top",
      legend.justification = "right",
      legend.margin = margin(b=.05, unit = "in")
  )

```

<br>

__Let's inspect the top up and down genes deeper !__

```{r ,fig.width=8, fig.height=10}
rbind(all_results_filt %>% 
        dplyr::filter(Gene_name %in% genes_list$up_genes) %>% 
        dplyr::arrange(desc(log2FoldChange)) %>% 
        head(25),
      all_results_filt %>% 
        dplyr::filter(Gene_name %in% genes_list$down_genes) %>% 
        dplyr::arrange(desc(log2FoldChange)) %>% 
        tail(35)) %>% 
  ggplot(aes(x= log2FoldChange, y= reorder(Gene_name, log2FoldChange)))+
  geom_bar(aes(fill = Gene_biotype),
           stat = "identity",
           width = 0.7)+
  scale_fill_manual(values = c("protein_coding"="darkred",
                               "lncRNA"="black"))+
  labs(title = "Biotypes of the top up and down genes",
       y = "")+
  my_general_theme()+
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.y = element_text(face = "bold")
  )

```



# Pathway enrichment analysis (PEA)

## Gene Ontology (GO) {.tabset .tabset-pills}

<br>

<div class="note">
<p class="p-120 p-bold p-purple p-underline">
NOTE:
</p>
<P class="p-bold">
For the downstream analysis, only `protein_coding` genes will be kept to ensure the significance of the resulting terms!
</p>
</div>

<br>

```{r}
## Params:
all_genes_protcoding <- all_results_filt %>% 
  dplyr::filter(Gene_biotype == "protein_coding") %>% 
  dplyr::pull(Gene_name)

```


### Results {.unnamed .unlisted}

```{r}
go.obj <- enrichGO(gene = all_genes_protcoding,
                   OrgDb = "org.Hs.eg.db",
                   keyType = "SYMBOL",
                   minGSSize = 10,
                   maxGSSize = 1000,
                   readable = T,
                   ont = "BP")

go.res <- go.obj@result %>%
  dplyr::filter(p.adjust < 0.05) %>%
  dplyr::arrange(p.adjust) %>% 
  dplyr::mutate(BgGenes = as.numeric(sub("/\\d+","",BgRatio)),
                RichFactor_ratio = paste0(Count,"/",as.numeric(sub("/\\d+","",BgRatio)))) %>% 
  dplyr::select(Description, RichFactor, p.adjust, GeneRatio, BgRatio, Count, BgGenes, RichFactor_ratio, geneID)

go.res %>% DT::datatable(options = list(pageLength = 5, scrollX = T))
```

<br>

[The top 30 most significant pathways enriched with these DEGs are]{.p-120 .p-bold .p-green}

```{r ,class.output="scroll-200-red"}
go.res %>% head(30) %>% dplyr::pull(Description)
```

<br>

[The top 30 pathway with the highest `RichFactor` (Count per pathway length ratio)]{.p-120 .p-bold .p-green}

```{r ,class.output="scroll-200-red"}
go.res %>% dplyr::arrange(desc(RichFactor)) %>%  head(30) %>% dplyr::pull(Description)
```

<br><br>

### Dotplot {.unnamed .unlisted}


```{r ,fig.height=10, fig.width=9}
terms <- go.res %>% head(40)

go.res %>% 
  dplyr::filter(Description %in% terms$Description) %>% 
  dplyr::mutate(Description = ifelse(nchar(Description) <= 50,
                                     Description,
                                     paste0(substr(Description,1,46), "...."))) %>% 
  ggplot(aes(x= RichFactor, y= forcats::fct_reorder(Description, RichFactor)))+
  geom_segment(aes(xend= 0, yend= Description))+
  geom_point(aes(color= p.adjust, size= Count))+
  geom_text(aes(label = RichFactor_ratio), size = 3, colour = "darkred", nudge_x = 0.025)+
  scale_color_viridis_c(guide = guide_colorbar(reverse = T))+
  scale_size_continuous(range = c(2,8))+
  labs(title = paste0("Top ", nrow(terms), " significant enriched pathways"),
       y = "")+
  xlim(c(0, (max(terms$RichFactor)+0.05)))+
  my_general_theme()+
  theme(
    axis.text.y = element_text(face = "bold")
  )
```

<br><br>

### Cnet plot {.unnamed .unlisted}

```{r ,fig.height=10, fig.width=12}

Net_plot(GOobject = go.obj,
         DEAres = all_results_filt,
         condition.name = "JK-H151",
         select.terms = terms$Description[1:4],
         net.layout = "nicely",
         edge.params = list(color = NULL,
                            alpha = .3,
                            width = .8),
         node.params = list(term.col = "black",
                            term.size = c(10,6),
                            gene.col = "logFC",
                            gene.shape =F,
                            gene.size = 6,
                            gene.alpha = .9),
         label.params = list(gene.col = c("navy", "black"),
                             gene.size = 3.5)) -> p1 ; p1$plot

```

<br>

```{r ,fig.height=7, fig.width=8}
term_list <- list()
for(i in unique(p1$netdata$Term)){
  term_list[[i]] <- p1$netdata$Gene[p1$netdata$Term == i]
}

Genes_comparison <- Display_Venn(Markers = term_list, 
                                 colpalette = NULL, 
                                 set.names = NULL, 
                                 set.name.size = 4,
                                 text.size = 4,
                                 Padding = 0.05)

Genes_comparison$plot
```

<br><br>

```{r ,fig.height=10, fig.width=12}

Net_plot(GOobject = go.obj,
         DEAres = all_results_filt,
         condition.name = "JK-H151",
         select.terms = terms$Description[5:8],
         net.layout = "nicely",
         edge.params = list(color = NULL,
                            alpha = .3,
                            width = .8),
         node.params = list(term.col = "black",
                            term.size = c(10,6),
                            gene.col = "logFC",
                            gene.shape =F,
                            gene.size = 6,
                            gene.alpha = .9),
         label.params = list(gene.col = c("navy", "black"),
                             gene.size = 3.5)) -> p2 ; p2$plot

```

<br>

```{r ,fig.height=7, fig.width=8}
term_list <- list()
for(i in unique(p2$netdata$Term)){
  term_list[[i]] <- p2$netdata$Gene[p2$netdata$Term == i]
}

Genes_comparison <- Display_Venn(Markers = term_list, 
                                 colpalette = NULL, 
                                 set.names = NULL, 
                                 set.name.size = 4,
                                 text.size = 4,
                                 Padding = 0.05)

Genes_comparison$plot
```

<br><br>

```{r ,fig.height=10, fig.width=12}

Net_plot(GOobject = go.obj,
         DEAres = all_results_filt,
         condition.name = "JK-H151",
         select.terms = terms$Description[9:12],
         net.layout = "nicely",
         edge.params = list(color = NULL,
                            alpha = .3,
                            width = .8),
         node.params = list(term.col = "black",
                            term.size = c(10,6),
                            gene.col = "logFC",
                            gene.shape =F,
                            gene.size = 6,
                            gene.alpha = .9),
         label.params = list(gene.col = c("navy", "black"),
                             gene.size = 3.5)) -> p3 ; p3$plot

```

<br>

```{r ,fig.height=7, fig.width=8}
term_list <- list()
for(i in unique(p3$netdata$Term)){
  term_list[[i]] <- p3$netdata$Gene[p3$netdata$Term == i]
}

Genes_comparison <- Display_Venn(Markers = term_list, 
                                 colpalette = NULL, 
                                 set.names = NULL, 
                                 set.name.size = 4,
                                 text.size = 4,
                                 Padding = 0.05)

Genes_comparison$plot
```

<br><br>

```{r ,fig.height=10, fig.width=12}

Net_plot(GOobject = go.obj,
         DEAres = all_results_filt,
         condition.name = "JK-H151",
         select.terms = terms$Description[13:16],
         net.layout = "nicely",
         edge.params = list(color = NULL,
                            alpha = .3,
                            width = .8),
         node.params = list(term.col = "black",
                            term.size = c(10,6),
                            gene.col = "logFC",
                            gene.shape =F,
                            gene.size = 6,
                            gene.alpha = .9),
         label.params = list(gene.col = c("navy", "black"),
                             gene.size = 3.5)) -> p4 ; p4$plot

```

<br>

```{r ,fig.height=7, fig.width=8}
term_list <- list()
for(i in unique(p4$netdata$Term)){
  term_list[[i]] <- p4$netdata$Gene[p4$netdata$Term == i]
}

Genes_comparison <- Display_Venn(Markers = term_list, 
                                 colpalette = NULL, 
                                 set.names = NULL, 
                                 set.name.size = 4,
                                 text.size = 4,
                                 Padding = 0.05)

Genes_comparison$plot
```


<br><br>

## Gene Set Enrichment Analysis (GSEA)

__GSEA__ is a functional class scoring method for enrichment analysis! It uses the whole set of genes ranked by the chosen scoring method: in our case, ranked by `p.adjust` and `logFC` values

- Higher scores represent the __most significant Up regulated genes__
- Lower scores represent the __most significant Down regulated genes__
- In between are the not significant genes

```{r echo=TRUE}
ranked_genes <- bitr(all_results$ensembl, 
                     fromType = "ENSEMBL",
                     toType = "ENTREZID",
                     OrgDb = "org.Hs.eg.db") %>% 
  dplyr::mutate(ensembl = ENSEMBL) %>% 
  inner_join(all_results, by = "ensembl") %>% 
  dplyr::select(ENSEMBL, ENTREZID, Gene_name, padj, log2FoldChange, Gene_biotype) %>% 
  dplyr::filter(Gene_biotype == "protein_coding",
                padj < 0.5,
                abs(log2FoldChange) >= 0.1) %>% 
  dplyr::mutate(padj = padj + 1e-300,
                rank = -log10(padj)*sign(log2FoldChange)) %>% 
  dplyr::group_by(ENTREZID) %>% 
  dplyr::slice_max(order_by = abs(rank), n = 1, with_ties = FALSE) %>% 
  ungroup() %>%
  dplyr::arrange(desc(rank))

ranked_genes <- setNames(ranked_genes$rank, ranked_genes$ENTREZID)
ranked_genes <- ranked_genes[intersect(names(ranked_genes), 
                                       keys(org.Hs.eg.db, keytype = "ENTREZID"))]

```


```{r}
gsea.obj <- gseGO(geneList = ranked_genes,
                  OrgDb = org.Hs.eg.db,
                  ont = "BP",
                  keyType = "ENTREZID",
                  minGSSize = 10,
                  maxGSSize = 1500,
                  pvalueCutoff = 0.05,
                  verbose = F,
                  nPermSimple = 10000) %>% 
  setReadable(OrgDb = 'org.Hs.eg.db', keyType = "ENTREZID")

gsea.result <- gsea.obj@result

if(nrow(gsea.result) > 0){
  gsea.result %>% datatable(options = list(pageLength = 10, scrollX = TRUE))
} else {
  print("No results found!")
}
```



















