---
title: "Jurkat project"
author: "Juba"
date: "`r Sys.Date()`"
output: 
  html_document:
    number_sections: no
    theme: flatly
    highlight: tango
    toc: yes
    toc_depth: 6
    toc_float:
      collapsed: false
      smooth_scroll: true
    df_print: paged 
    css: "style.css"
---

```{css ,echo=FALSE}

```


```{r setup, include=FALSE, message = FALSE, warning = FALSE, echo = FALSE}
## Chunks configuration :
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)

## Directories :
out_fig_explore <- "out_figures/exploration/"
out_fig_dea <- "out_figures/dea/"
out_obj <- "out_objects/"

## Import packages :
source("scripts/packages.R") 
```


# Project description

We are studying the effect of an inhibitor of the cGAS-STING signaling pathway __H151__ on T-ALL model cell line __Jurkat__.

The biological experiment revealed that H151 causes cell death, so the pathway is important 
for survival of T-ALL.

[We want to explore the differential expressed genes between normal condition and by inhibiting the pathway. Thus we are going to perform a differential expression analysis (DEA) followed by a pathway enrichment analysis (PEA)]{.p-purple .p-bold}

__Following a basic RNA-seq pipeline analysis__



# Importing data {.tabset .tabset-fade}

```{r import-data}
## ---- Import all the data : ----
############################# /
all_data <- read_xlsx(path = "data/S22254_alldata.xlsx",
                      sheet = "S22254_alldata")

colnames(all_data) <- gsub(" ", "_", colnames(all_data))



## ---- Extract raw counts : ----
############################ /
raw_counts <- all_data[,1:7] %>% 
  column_to_rownames("Ensembl_Gene_ID") %>% 
  dplyr::rename(LXGP1 = "LXGP1_(raw)",
                LXGP2 = "LXGP2_(raw)",
                LXGP3 = "LXGP3_(raw)",
                LXGP4 = "LXGP4_(raw)",
                LXGP5 = "LXGP5_(raw)",
                LXGP6 = "LXGP6_(raw)")



## ---- Extract meta data : ----
########################### /
metadata <- all_data[,c(1,27:31)] %>% 
  column_to_rownames("Ensembl_Gene_ID") %>% 
  mutate(Gene_Length = abs(Start_gene_position - End_gene_position))



## ---- Extract experimental data : ----
################################### /
experimental_data <- read_xlsx(path = "data/Table_Experimental conditions.xlsx") %>% 
  select(c(1,3,5,7,9)) %>% 
  drop_na() %>% 
  column_to_rownames(var = "Code") %>% 
  dplyr::rename(Concentration = `Concentration (ng/µl)`,
                Volume = `Volume (µl)`) %>% 
  head(6) %>% 
  mutate(Condition = c(rep("Jurkat_ct",3), rep("Jurkat_H151",3)),
         Replicat = rep(c("1","2","3"), times = 2))

## 4: saving objects |||
write_xlsx(raw_counts, path = paste0(out_obj, "rawcounts.xlsx"))
write_xlsx(metadata, path = paste0(out_obj, "metadata.xlsx"))
write_xlsx(experimental_data, path = paste0(out_obj, "experimental_data.xlsx"))
```

## Raw counts {.unnumbered .unlisted}

```{r ,rows.print=5}
raw_counts
```

## Metadata {.unnumbered .unlisted}

```{r ,rows.print=5}
metadata
```

## Experimental data {.unnumbered .unlisted}

```{r ,rows.print=5}
experimental_data
```




# Data exploration

First let's view the distribution of the different bio types we have in our data : 

```{r biotypes-plot, fig.width=8, fig.height=6}
metadata %>% 
  group_by(Gene_biotype) %>% 
  summarise(counts = n()) %>% 
  filter(counts > 10) %>% 
  ggplot()+
  geom_bar(aes(x= counts,
               y= Gene_biotype,
               fill = ifelse(Gene_biotype == "protein_coding", "darkcyan", "gray")),
           stat = "identity",
           show.legend = F,
           width = 0.8)+
  theme_bw()+
  labs(title = "Gene biotypes distributions",
       y = "",
       fill = "",
       x = "Counts")+
  theme(plot.title = element_text(size = 14,
                                  colour = "darkcyan",
                                  face = "bold"),
        plot.background = element_rect(colour = "whitesmoke", 
                                       fill = "whitesmoke", 
                                       linewidth = 1),
        axis.text = element_text(size = 11),
        axis.title = element_text(size = 12, 
                                  colour = "darkcyan", 
                                  face = "bold"))+
  scale_fill_manual(values = c("darkcyan"="darkcyan", "gray"="gray60"))+
  xlim(c(0,17000))+
  geom_text(aes(x= counts,
                y= Gene_biotype,
                label = counts),
            nudge_x = 700,
            size = 3.5)
```

[In the downstream analysis (DEA), we'll be focusing on the top 2 biotypes (`protein_coding` and `lncRNA`)]{.p-120 .p-red .p-bold}

```{r filtering-data}
to_keep <- rownames(metadata[which(metadata$Gene_biotype %in% c("protein_coding","lncRNA")),])
raw_counts_filt <- raw_counts[to_keep,]

rm(to_keep)
```


<br>

Now to understand the global gene expression landscape and to assess the quality control of our data, we need to perform a dimentionality reduction analysis : [Principal component analysis (PCA)]{.p-bold .p-purple}

The PCA :

- Assumes homoscedasticity (equal variance) between the samples
- Assumes roughly normal distribution

So we need to perform some data transformation first on `raw counts`.

We'll be using a [Variance Stabilizing Transformation (VST)]{.p-bold .p-purple} from the `DESeq2` package. And This will :

- Normalize for library size (`DESeq2` computes size factors using a `median-of-ratios` method)
- Use a __negative binomial__ model to estimate dispersion
- Apply a variance-stabilizing mathematical transformation

```{r pca-fit}
## 1: build DESeq2 structure
dds <- DESeqDataSetFromMatrix(countData = raw_counts_filt,
                              colData = experimental_data,
                              design = ~ Condition)

## 2: apply vst
vsd <- vst(dds, blind = T)
#### blind = TRUE (gives a neutral VST - not influenced by the design 'condition')

## 3: extract vst matrix 
vst_mat <- assay(vsd)

## 4: run PCA using prcomp()
pca <- prcomp(t(vst_mat), scale. = TRUE)
```

```{r pca-plot}
autoplot(pca, 
         data = experimental_data, 
         colour = "Condition",
         size = 4)+
  labs(title = "PCA of Jurkat_ct vs Jurkat_H151")+
  theme_bw()+
  scale_colour_manual(values = c("Jurkat_H151"="brown", "Jurkat_ct"="black"))+
  theme(
    plot.title = element_text(size = 14,
                              colour = "darkcyan",
                              face = "bold"),
    plot.background = element_rect(colour = "whitesmoke",
                                   fill = "whitesmoke",
                                   linewidth = 1),
    legend.key = element_rect(colour = "whitesmoke",
                              fill = "whitesmoke",
                              linewidth = 1),
    legend.background = element_rect(colour = "whitesmoke",
                                     fill = "whitesmoke",
                                     linewidth = 1),
    legend.title = element_text(colour = "darkcyan",
                              size = 12,
                              face = "bold")
  )
```












